### ISSUE TITLE
High amount of garbage created in BasePutOperation.afterRun()

### ISSUE THREAD

**Issue by stffrdhrn** [hazelcast/hazelcast#17206]
**Please describe the problem you are trying to solve**
TL/DR; `BasePutOperation.afterRun()` is creating a lot of garbage.

We have deployed hazelcast in a high throughput environment where we like to see low GC overhead.  We have several eventListeners registered with a map at any time and only 1 publisher.  In our environment we were seeing >2 million updates total to ~1.5million records in 10 minutes for a map hosted on 2 nodes.  

We observed a high amount of garbage created with the normal setup.  Below is a summary of the observed TLAB allocations for a ~10 minutes period with java flight recorder.

In `AbstractPartitionMessageTask.processMessage()`
  - in `BasePutOperation`:
     - During `afterRun()`   `250GB`  _(for de-serialization of objects when running EntityListener predicates)_
  - in `DefaultRecordStore`:
     - During `saveIndex()`   -  `130GB`  _(for de-serialization of objects storing to cache)_
     - During `createRecord()/updateRecord()`  -  `99MB`

After that I configured my IMap cache with InMemoryFormat to `OBJECT` and CacheDeserializedValues to `NEVER`.   The flight recorder TLAB allocations after that show:

In `AbstractPartitionMessageTask.processMessage()`
  - in `BasePutOperation`:
     - During `afterRun()`   `100GB`  _(still for de-serialization of objects when running EntityListener predicates)_
  - in `DefaultRecordStore`:
     - During `createRecord()/updateRecord()`  -  `16GB` _(for de-serialization of objects storing to map)_
     - During `saveIndex()`   -  `1GB`

I chose to use InMemoryFormat to `OBJECT` as our objects:
  1. Already use a compact in memory format
  2. We need to often query all records and de-serialization overhead causes a large amount of garbage to be created

As expected the amount of objects created during cache updates decreased and the number of objects created for records increased.  However, there is still a large amount of data created during `afterRun()`.

In [BasePutOperation](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/operation/BasePutOperation.java) `afterRun()` the value sent to the eventPublisher uses the serialized `dataValue`, unless we use `isPostProcessing()`, this means the eventPublisher will always need to deserialize the object again.

```
        Object value = isPostProcessing(recordStore)
                ? recordStore.getRecord(dataKey).getValue() : dataValue;
        ...
        mapEventPublisher.publishEvent(getCallerAddress(), name, getEventType(),
                dataKey, oldValue, value);
```

**Please describe the desired behavior**
Ideally hazelcast could recognize that its more efficient to pass the de-serialized object directly
to the event publisher.  It could be as simple as:

```
-        Object value = isPostProcessing(recordStore)
+        Object value = (OBJECT == inMemoryFormat || isPostProcessing(recordStore))
                ? recordStore.getRecord(dataKey).getValue() : dataValue;
        ...
        mapEventPublisher.publishEvent(getCallerAddress(), name, getEventType(),
                dataKey, oldValue, value);
```

**Describe alternatives you've considered**
The de-serialized object record will be sent to the event publisher if the call to `isPostProcessing(recordStore)` returns true. This will be done if:

1. We have setup a PostProcessingMapStore
2. We have an interceptor registered with the map

Both methods can work to decrease the GC overhead.

## Workaround 1: Using a empty PostProcessingMapStore
I have worked around this by creating a `NullMapStore` implementation of `MapStore` that overrides `PostProcessingMapStore`.  This tricks hazelcast to use the records `OBJECT` format data value.

## Workaround 2: Using interceptors
Trying to use interceptors does not work for us as garbage is created while calling `interceptPut()` in `DefaultRecordStore.putInternal()` when calling [MapServiceContextImpl.interceptPut()](https://github.com/hazelcast/hazelcast/blob/44b2b6e4d3a7cdb6c3034fba4d256a47f8036ec0/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java#L595).  That could be fixed too, by either avoiding unimplemented calls to `interceptPut()`, or doing the deserialization once and not discarding it each time.

**Comment by ahmetmircik** [comment]
Hi @stffrdhrn,

Thanks for raising this. 

You are right, i can see here the main issue is even you have OBJECT in-memory-format, de-serialization is still keep happening and it is causing garbage creation:
```
 Object value = isPostProcessing(recordStore)
                ? recordStore.getRecord(dataKey).getValue() : dataValue;
        ...
        mapEventPublisher.publishEvent(getCallerAddress(), name, getEventType(),
                dataKey, oldValue, value);
```

At first glance, maybe we can also pass de-serialized value besides serialized one to `mapEventPublisher.publishEvent` method and use wherever it appropriate.

Do you want to work on this and send a fixer PR?

**Comment by stffrdhrn** [comment]
Sure, let me work on it.

**Comment by stffrdhrn** [comment]
A pull request is created for this: https://github.com/hazelcast/hazelcast/pull/17216

**Comment by mmedenjak** [comment]
Moving the issue and PR into backlog - can be resurrected after the 4.2 release by someone from the team.
